pkg/catalogv2/oci/oci.go

```diff
@@ -128,7 +128,9 @@ func GenerateIndex(ociClient *Client, URL string, credentialSecret *corev1.Secre
 		for i := len(tags) - 1; i >= 0; i-- {
 			existingTags[tags[i]] = true
 			// Check if the tag is a valid semver version or not. If yes, then proceed.
-			semverTag, err := version.NewVersion(tags[i])
+			// Change underscore (_) back to plus (+) same as Helm does
+			// See https://github.com/helm/helm/issues/10166
+			semverTag, err := version.NewVersion(strings.ReplaceAll(tags[i], "_", "+"))
```

Comment:
This should only replace the first instance to a plus. Any subsequent ones should be a period or a hyphen.

This is because plus is only a valid character to delineate between main SemVer and the build metadata. So it can on be used once.

Reply:
You're right, are you suggesting this due to performance of using string.ReplaceAll?

See [Helm codebase comment](https://github.com/helm/helm/blob/7f83ea5e5ecee756bbc62783bd6ad2ea1e2dc6bf/pkg/registry/client.go#L49-L52) and actual [code which does the conversion](https://github.com/helm/helm/blob/7f83ea5e5ecee756bbc62783bd6ad2ea1e2dc6bf/pkg/registry/client.go#L745) when pulling from the registry. When pushing to registry, Helm converts all + to _ and when reading from registry it converts all _ to +.

So my change does exactly the same thing on Rancher side, when reading from the OCI registry. If the version is not valid after the conversion, the error is returned and the tag is skipped - same as it happens right now when there is _ which causes the issue.

Reply:
> Helm converts all + to _ and when reading from registry it converts all _ to +

oh I didn't realize that part.

I also realize - in hindsight - that because `_` is not a valid character to use within `<build>` or  `<pre-release>` of semver that my concern is moot. I was considering potential edge cases this might cause (if versions already had `_` in them), but completely overlooked that wouldn't be a valid version to start with. 🤦